Automatically set 0 to deregistered subnets

shibshib opened the pull request:
This will ensure that when subnets are deregistered, its weights are automagically set to 0.

Commit:
Adding a note on installing on cloud

Commit:
Added call  to set weights to 0 for deregged subnets

shibshib committed:
Fixed logic to remove weights

Commit:
Fixed numbering on commits

Commit:
Added unit test for subnet pruning and weight reset mechanism

shibshib committed:
Merge branch 'main' into subnet_dereg_automatic_weighing

sam0x17 commented on pallets/subtensor/src/root.rs:
```diff
@@ -821,9 +822,29 @@ impl<T: Config> Pallet<T> {
+                } else {
+                    // For all other entries, just copy them to the new vector.
+                    modified_weights.push((*subnet_id, *weight));
+                }
```
This is inefficient for a few reasons:
1. If we use `Vec::new()` every time, but we actually know the final size of the Vec, we are relying on the auto-resizing mechanics of `Vec` when we could just make the whole allocation up-front, resulting in a few intermediate needless allocations
2. It turns out though, we are better off just cloning the original vec, since we are just taking the original `weights_i` vec and making some minor modifications. `clone()` for `Vec` uses the `memcpy` system call, which is much faster than manually copying via a for loop
3. Then in the actual for loop, we simply set the weight to zero for entries where the `subnet_id` matches `netuid`:

```suggestion
            let mut modified_weights = weights_i.clone();
            // Iterate over each weight entry to potentially update it.
            for (subnet_id, weight) in modified_weights.iter_mut() {
                if subnet_id == &netuid {
                    // If the condition matches, modify the weight
                    *weight = 0; // Set weight to 0 for the matching subnet_id.
                }
```

that said, I do wonder if this can be made even more efficient. If we could know ahead of time the index/indices where `subnet_id` is equal to `netuid`, then we could avoid looping over the whole vec, which would be much more efficient. What is the structure of `weights_i` going into this loop? I imagine we would only be stuck with this setup if `weights_i` has no pattern to it at all in terms of the `subnet_id`, is unsorted, and has an unknown number of duplicates. For example, if we know minimally that a given `subnet_id` only appears once, we can break early at least. If we know a `subnet_id` only appears once, and at a specific position, we can skip the loop entirely and index into that position, etc

sam0x17 replied:
on the flipside, if it really is unsorted and with an unknown number of duplicates, the original code with `Vec::with_capacity(weights_i.len())` would be more efficient than what I have written or what was previously there ;)

sam0x17 reviewed (commented):
Looks good in general, one loop I think could be improved significantly

shibshib committed:
Update pallets/subtensor/src/root.rs

Updated weight setting to directly use .clone() for much faster copying than the old loop style.

Co-authored-by: Sam Johnson <sam@durosoft.com>

shibshib commented:
Thanks @sam0x17 this was super helpful, updated the loop after testing.

sam0x17 commented:
@shibshib did you have any thoughts regarding what I was saying about "is there anything we can do to avoid looping over the whole vec" -- I don't have enough context to know the topology of the `weights_i` vec but I'm curious what the details are here, like can an id be repeated, is the order random, etc

shibshib committed:
Merge remote-tracking branch 'origin/weights_min_stake' into subnet_dereg_automatic_weighing

shibshib commented:
> @shibshib did you have any thoughts regarding what I was saying about "is there anything we can do to avoid looping over the whole vec" -- I don't have enough context to know the topology of the `weights_i` vec but I'm curious what the details are here, like can an id be repeated, is the order random, etc

We unfortunately have to loop through the whole vec. The vector itself contains the weights set by validators for all subnets. If a subnet's weight must be reset to 0, it has to be reset by every validator as well. Hence, it's hard to optimize that further, I don't think. Open to suggestions though!

sam0x17 commented:
should be good as-is then 👍🏻

unconst reviewed (approved):
LGTM

Commit:
Fixed accidental addition of new event

shibshib committed:
Merge branch 'subnet_dereg_automatic_weighing' of https://github.com/opentensor/subtensor into subnet_dereg_automatic_weighing

Commit:
Added network removed event upon network removal when registering a new network

shibshib committed:
Merge branch 'subnet_dereg_automatic_weighing' of https://github.com/opentensor/subtensor into subnet_dereg_automatic_weighing

shibshib committed:
Added some PR reviewer comment suggestions

shibshib committed:
Added some PR reviewer comment suggestions

shibshib committed:
Merge branch 'main' of https://github.com/opentensor/subtensor into subnet_dereg_automatic_weighing

shibshib committed:
Fixed spec_version
